spikeextractors/extractors/nwbextractors/nwbextractors.py

```diff
@@ -367,81 +340,76 @@ def add_electrode_groups(recording: se.RecordingExtractor, nwbfile=None,
-                device = nwbfile.devices[device_names[0]]
-
-            electrode_group_kwargs = fill_kwargs_from_defaults(defaults)
-            electrode_group_kwargs.update({'device': device})
-
-            if 'group' in recording.get_shared_channel_property_names():
-                RX_groups_names = np.unique(recording.get_channel_groups()).tolist()
-                for grp_name in RX_groups_names:
-                    electrode_group_kwargs.update(
-                        {'name': str(grp_name)})  # Over-write default name with internal value
-                    nwbfile.create_electrode_group(**electrode_group_kwargs)
-            else:
-                nwbfile.add_electrode_groups(**electrode_group_kwargs)
-                print("Warning: No electrode group metadata provided, and no"
-                      " internal group properties set. Generating default device!")
+            metadata = dict(
+                Ecephys=dict(
+                    ElectrodeGroup=[]
+                )
+            )
+        if 'group' not in recording.get_shared_channel_property_names():
+            metadata['Ecephys']['ElectrodeGroup'] = [dict(defaults)]
+        assert all([isinstance(x, dict) for x in metadata['Ecephys']['ElectrodeGroup']]), \
+            "Expected metadata['Ecephys']['ElectrodeGroup'] to be a list of dictionaries!"
+
+        for grp in metadata['Ecephys']['ElectrodeGroup']:
+            device_name = grp.get('device_name', defaults['device_name'])
+            if grp.get('name', defaults['name']) not in nwbfile.electrode_groups:
+                if device_name not in nwbfile.devices:
+                    new_device = dict(
+                        Ecephys=dict(
+                            Device=dict(
+                                name=device_name
+                            )
+                        )
+                     )
+                    se.NwbRecordingExtractor.add_devices(recording, nwbfile, metadata=new_device)
+                    print("Warning: device name not detected in attempted link to electrode group! "
+                          "Automatically generating.")
+                electrode_group_kwargs = dict(defaults, **grp)
+                electrode_group_kwargs.pop('device_name')
+                electrode_group_kwargs.update(dict(device=nwbfile.devices[device_name]))
+                nwbfile.create_electrode_group(**electrode_group_kwargs)
+
+        if 'group' in recording.get_shared_channel_property_names():
+            RX_groups_names = np.unique(recording.get_channel_groups()).tolist()
+
+            if len(nwbfile.devices) == 0:
+                se.NwbRecordingExtractor.add_devices(recording, nwbfile)
+            elif len(nwbfile.devices) > 1:
+                print("Warning: more than one device found when adding electrode group via channel properties: "
+                      "using first added device. To use a different device, indicate it the metadata argument.")
```

Comment:
would be nice if this warning message
1. Used `warnings.warn`
2. Indicated the name of the device that was being used. The "first" key may or may not be deterministic, depending on the version of python being used.

Reply:
Gotcha, no longer referring to it as "first" device then, just by whatever its name is (see the other PR for changes).